### PR TITLE
Add pseudo-inverse network

### PR DESCRIPTION
**Pull request by neale**
* added pinverse_network.py to alf/networks
    pinverse_network contains the network definition and forward method for a network that should in principle output an inverse Jacobian vector product.
    This network is used for training GPVI
* added pinverse_algorithm to alf/algorithms/generator.py
   the pinverse_algorithm specifies a training step for the pinverse network.

* added configuration options to alf/algorithms/generator.py for the pinverse network, as well as an option for ``functional gradient``.
    as of now, the configurations proposed in this PR to the generator __init__ function do nothing downstream.
    But they will be used in the subsequent commit.

### CODE REVIEW

File: alf/algorithms/generator.py [hunk 1]
```diff
@@ -115,6 +115,85 @@ def predict_step(self, inputs, state=None, requires_jac_diag=False):
...
+            net (Network): network for predicting outputs from inputs.
+                If None, a default ReluMLP with hidden_layers will be created
+            optimizer (torch.optim.Optimizer): (optional) optimizer for training.
+            name (str): name of this CriticAlgorithm.
+            eps_shape (list or None): shape of eps to estimate. In general ``svgd3``
+                takes an eps input of shape [B', B, k]. ``minmax`` takes an eps
+                of shape [B, k, d]. 
+                if ``eps_shape`` is None, then only :math:'J^{-1}' is estimated, 
+                and eps is not an input to the forward pass
+        """
```
Comment (runjerry) [inline alf/algorithms/generator.py:159]: The description of arguments is incorrect.

File: alf/algorithms/generator.py [hunk 2]
```diff
@@ -282,6 +376,32 @@ def __init__(self,
             else:
                 raise ValueError("Unsupported par_vi method: %s" % par_vi)
 
+            if functional_gradient is not None:
+                if functional_gradient == 'rkhs':
+                    self._grad_func = self._rkhs_func_grad
+                    if force_fullrank:
```
Comment (runjerry) [inline alf/algorithms/generator.py:382]: force_fullrank is not defined?

File: alf/algorithms/generator.py [hunk 3]
```diff
@@ -282,6 +376,32 @@ def __init__(self,
...
+                        self._eps_dim = output_dim
+                    else:
+                        self._eps_dim = noise_dim
+                else:
+                    raise ValueError(
+                        'functional gradient only supports ``rkhs``')
+
+                if noise_dim == output_dim:
+                    force_fullrank = False
```
Comment (runjerry) [inline alf/algorithms/generator.py:391]: force_fullrank is not used anywhere afterward?

**Review by runjerry — CHANGES_REQUESTED** [review]
I think this PR is confusing given that some changes seem to me just placeholders for PR #932. Maybe just merge it into #932 instead.
